src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 and BSD-3-Clause
+
+//! System device management.
+//!
+//! [DeviceManager](struct.DeviceManager.html) responds to manage all devices
+//! of virtual machine, store basic device information like name and
+//! parent bus, register IO resources callback, unregister devices and help
+//! VM IO exit handling.
+
+extern crate vm_allocator;
+
+use self::vm_allocator::{Error as AllocatorError, SystemAllocator};
+use crate::device::*;
+use std::cmp::{Ord, Ordering, PartialEq, PartialOrd};
+use std::collections::btree_map::BTreeMap;
+use std::collections::HashMap;
+use std::result;
+use std::sync::{Arc, Mutex};
+use vm_memory::{Address, GuestAddress, GuestUsize};
+
+/// Guest physical address and size pair to describe a range.
+#[derive(Eq, Debug, Copy, Clone)]
+pub struct Range(pub GuestAddress, pub GuestUsize);
+
+impl PartialEq for Range {
+    fn eq(&self, other: &Range) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Ord for Range {
+    fn cmp(&self, other: &Range) -> Ordering {
+        self.0.cmp(&other.0)
+    }
+}
+
+impl PartialOrd for Range {
+    fn partial_cmp(&self, other: &Range) -> Option<Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+/// Error type for `DeviceManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The insertion failed because the new device overlapped with an old device.
+    Overlap,
+    /// PIO request is none.
+    NonePIORequest,
+    /// The insertion failed because device already exists.
+    Exist,
+    /// The removing fails because the device doesn't exist.
+    NonExist,
+    /// Irq allocation failed.
+    IrqAllocate(AllocatorError),
+    /// Instance id allocation failed.
+    InstanceIdAllocate(AllocatorError),
+    /// Address allocation failed.
+    AddressAllocate(AllocatorError),
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+/// System device manager serving for all devices management and VM exit handling.
+pub struct DeviceManager {
+    /// System allocator reference.
+    resource: Arc<Mutex<SystemAllocator>>,
+    /// Devices information mapped by instance id.
+    devices: HashMap<u32, DeviceDescriptor>,
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
```

Comment:
I have strong concern about the design decision to protect Device object with Arc<Mutex<>> at the DeviceManager layer. That means all accesses to the Device object will be serialized by the device manager, and virtio multi-queues features will be effectively disabled by this design. Actually we should let device backend to decide how to protect its fields.
So mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>> should be
mmio_bus: BTreeMap<Range, Arc<dyn Device>> instead.
And a DummyDevice may be implemented as :
struct DummyDevice {
     ro_field: u32,
    rw_field: RwLock<u32>,
    rw_field2: Mutex<u32>,
};

Reply:
> virtio multi-queues features will be effectively disabled by this design

Could you please elaborate on that?

Reply:
@sameo I guess it means multiple queues might notify vmm that some buffers should be handled at the same time, with different threads?

Reply:
Yes. The device may be accessed by different vcpus concurrently, but current design of the device manager enforces serial accesses due to the Arc<Mutex<dyn Device>>.

Reply:
There might have a problem: we should assume and limit the device instance to be only one pointer. Because when we do `dev.set_resources()` which needs mutable self, `Arc<> provides related APIs like `get_mut()` and `make_mut()` both assuming there's one pointer, or else it would clone the inner value or panic. So we should also limit the VMM not clone pointer before register_device. Use get_mut() to use panic to protect unique inner value.
```
    pub fn register_device(
        &mut self,
        dev: Arc<dyn Device>,
         ...
    ) -> Result<()> {
             ...
        dev.get_mut().set_resources(resources);
             ...
    }
```

Reply:
As discussed with @sameo , the key point here is that Arc<> limits the pointer numbers once we need direct the inner value to be mutable referece. But with Mutex<>, MutexGuard help impl DerefMut trait so things become easier. 

I think we can set all functions of `Device` trait to be `&self` instead of `&mut self`, so that DeviceManager don't need care mutable inner device and let device structure decide which members should be `Mutex` protection to solve this problem. Let's update the codes to discuss more.

Reply:
Yes, it means to design the device trait with interior mutability.